components/Chips/src/MDCChipField.m

```diff
@@ -0,0 +1,573 @@
+                                                   chipFieldSize:self.bounds.size];
+
+  [self updateTextFieldPlaceholderText];
+}
+
+- (void)updateTextFieldPlaceholderText {
+  // Place holder label should be hidden if showPlaceholderWithChips is NO and there are chips.
+  // MDCTextField sets the placeholderLabel opacity to 0 if the text field has no text.
+  self.textField.placeholderLabel.hidden =
+      (!self.showPlaceholderWithChips && self.chips.count > 0) || ![self isTextFieldEmpty];
+}
+
++ (UIFont *)textFieldFont {
+  return [UIFont systemFontOfSize:[UIFont systemFontSize]];
+}
+
+- (CGSize)intrinsicContentSize {
+  CGFloat minWidth =
+      MAX(self.minTextFieldWidth + self.contentEdgeInsets.left + self.contentEdgeInsets.right,
+          self.bounds.size.width);
+  return [self sizeThatFits:CGSizeMake(minWidth, CGFLOAT_MAX)];
+}
+
+- (CGSize)sizeThatFits:(CGSize)size {
+  NSArray *chipFrames = [self chipFramesForSize:size];
+  CGRect lastChipFrame = [chipFrames.lastObject CGRectValue];
+  CGRect textFieldFrame = [self frameForTextFieldForLastChipFrame:lastChipFrame
+                                                    chipFieldSize:size];
+
+  // Calculate the required size off the text field.
+  // To properly apply bottom inset: Calculate what would be the height if there were a chip
+  // instead of the text field. Then add the bottom inset.
+  CGFloat height = CGRectGetMaxY(textFieldFrame) + self.contentEdgeInsets.bottom +
+      (self.chipHeight - textFieldFrame.size.height) / 2;
+  CGFloat width = MAX(size.width, self.minTextFieldWidth);
+
+  return CGSizeMake(width, height);
+}
+
+- (void)clearTextInput {
+  self.textField.text = kMDCEmptyTextString;
+  [self updateTextFieldPlaceholderText];
+}
+
+- (void)setChips:(NSArray<MDCChipView *> *)chips {
+  if ([_chips isEqual:chips]) {
+    return;
+  }
+
+  for (MDCChipView *chip in _chips) {
+    [self removeChipSubview:chip];
+  }
+
+  _chips = [chips mutableCopy];
+  for (MDCChipView *chip in _chips) {
+    [self addChipSubview:chip];
+  }
+  [self notifyDelegateIfSizeNeedsToBeUpdated];
+  [self invalidateIntrinsicContentSize];
+  [self setNeedsLayout];
+}
+
+- (BOOL)becomeFirstResponder {
+  return [self.textField becomeFirstResponder];
+}
+
+- (BOOL)resignFirstResponder {
+  [super resignFirstResponder];
+  return [self.textField resignFirstResponder];
+}
+
+- (void)addChip:(MDCChipView *)chip {
+  // Note that |chipField:shouldAddChip| is only called in |createNewChipFromInput| when it is
+  // necessary to restrict chip creation based on input text generated in the user interface.
+  // Clients calling |addChip| directly programmatically are expected to handle such restrictions
+  // themselves rather than using |chipField:shouldAddChip| to prevent chips from being added.
+  [_chips addObject:chip];
+  if ([self.delegate respondsToSelector:@selector(chipField:didAddChip:)]) {
+    [self.delegate chipField:self didAddChip:chip];
+  }
+  [self addChipSubview:chip];
```

Comment:
Consider a slightly different ordering here? I'm thinking that `_chips` should be up-to-date once the view is added to the hierarchy. The delegate should only be called once the field is "ready".

```objectivec
[_chips addObject:chip];
[self addChipSubview:chip];
[delegate callDelegate];
```

Also reordering the remove as well?

```objectivec
[self removeChipSubview:chip];
[_chips removeObject:chip];
[delegate callDelegate];
```

Reply:
Fixed addChip behavior. I think remove is already correct if our goal is for _chips to be up to date once the view hierarchy changes.

Reply:
Sounds good. Thanks!